### DISCUSSION TITLE
Pentair iChlor 30 Control

### DISCUSSION THREAD

**Discussion by VincentLombardi** [tagyoureit/nodejs-poolController#780] [General]
Hi,

I have just gotten NJSPC running, and I am having issues controlling my chlorinator. I have a ET P8, with an external load center for my chlorinator. The chlorinator I have is a [Pentair iChlor 30](https://www.pentair.com/en-us/products/residential/pool-spa-equipment/pool-water-treatment/ichlor_salt_chlorinegenerator.html). It uses the same external load center as the Pentair IntelliChlors. I know this because I had an IntelliChlor 20 for many years before it died and was replaced with the iChlor 30 using the same load center.

I have connected the chlorinator load center over RS485 to the ET, and NJSPC to the ET as well. On the ET, I can see the salt level reported, and using the OCP screen I can set the pool output % level. In NJSPC, I can see the salt PPM, as well as the output % set by the OCP, but when adjusting the output % in NJSPC, it does not change on the chlorinator or on the OCP. Originally, when the chlorinator was auto-discovered, I had not set a model type in NJSPC and when attempting to change the output % I got an error like [this](https://github.com/tagyoureit/nodejs-poolController/discussions/758) thread. I went into settings, and could not find the iChlor 30 as a model listed, so I chose the IntelliChlor 20. Now I do not get the error, but as a mentioned, I cannot control the output % from NJSPC.

If you need a replay of the RS485 communication let me know, I can record me sending commands from NJSPC as well as setting the output % on the OCP so we can see the data being sent.

Thanks!

**Comment by rstrouse** [comment]
Yes we will need a replay of you attempting to set the percentage from both places.  iChlor30 isn't listed use IntelliChlor 20.

**Reply by tagyoureit** [reply]
There is another packet capture linked from this thread:  https://github.com/tagyoureit/nodejs-poolController/issues/157

At some point the iChlor was reporting the water temp but I could never nail down when or replicate it.  Maybe we will have more luck this time, but otherwise the comms should be similar.

**Reply by VincentLombardi** [reply]
Okay, I have recorded some data.

A couple notes about my setup that might be relevant:
1. I run NJSPC and DashPannel on a VM in my house, they are both running as docker containers
2. I connect the OCP to these docker containers using a RS485 to Ethernet adapter. On the RS485 bus, there is the OCP, iChlor30, and RS485 to Ethernet adapter that acts as a TCP server. NJSPC connects as a client to read the RS485 bus. It is a usr-dr302.
3. Even before setting up NJSPC I never used my ET system, hence no schedules are circuits adequately set up. I just added two intellivalves and I am slowly building out my system.
4. My Pi with REM is currently unplugged so might be throwing errors

@tagyoureit Yes you are right, if I put the iChlor in "Diagnostic Mode", it reports Total Cell Hours, Water Temperature, Salinty (PPM), and Cell Voltage on its screen.

Onto the captures. In the first one, I waited for the config to reload and once it did (not sure if this is a bug or not) I had to go to settings>controller>comms>Prmary Port> and click "Save Port" for the RS485 communication to start. Once the config began to load, I have another issue where my time at the top flashed between EST (my timezone) and GMT. I notice when the config reloads the "Clock source" changes from Manual to Server. I should probably just set the container time to EST and I assume it will fix this. Anyways, changed it back to my TZ  using Manual as the Clock Source (around 1:00pm). Then went onto the OCP and adjusted the Pool Output %, waited a few seconds, and saw it update on the iChlor 30s screen. No update in NJSPC though, as it was showing the chlorinator as off.

[OCP Control.zip](https://github.com/tagyoureit/nodejs-poolController/files/11680349/OCP.Control.zip)

In the second capture, same first setup, click "Save Port", fix time. Then, using dashPannel, adjusted the Pool Setpoint %. Wait a few seconds, it updates on NJSPC but not on the iChlor or OCP. Then I adjusted it a bit more through NJSPC, same result. And finally, adjusted output % through the OCP, watched it update on the iChlor but not on dashPannel.

[NJSPC + OCP Control.zip](https://github.com/tagyoureit/nodejs-poolController/files/11680351/NJSPC.%2B.OCP.Control.zip)

Let me know if this is sufficient. I also get this weird issue sometimes where I get duplicate but null bodies, sometimes there is 1 as pictured below, and sometimes there are 4 null bodies. This is another issue but it's probably due to my setup using Docker, RS485 to ethernet, etc. I also seemed to lose some of my circuits/feature settings upon config reload. I had gone into Circuits>Aux-Circuits and set Pool to "Dont Stop" (I want the chlorinator on 24/7 as my pump runs 24/7 currently until I get my ET/NJSPC setup, chlorinator will drop to 0% unless the filter pump/pool circuit is on. Pump is a SuperFlo VS that is not actually controlled by the ET8/NJSPC yet but I have the cable and relays to get it setup) but upon reset it went back to the default 8 hours, just some weird issues.

<img width="615" alt="Screenshot 2023-06-07 125924" src="https://github.com/tagyoureit/nodejs-poolController/assets/33071399/94c3ebe1-dd7c-49c5-8d31-3061727588b1">

**Reply by rstrouse** [reply]
Yeah that is IntelliConnect not ET8.  The issue is that the ET8 isn't accepting the request to change the output.

**Reply by VincentLombardi** [reply]
> Yeah that is IntelliConnect not ET8. The issue is that the ET8 isn't accepting the request to change the output.

Does this mean the iChlor 30 will not work properly with the ET8/NJSPC? I only got it because my IntelliChlor 20 died and this is what was recommended to me at the pool store. I was much more naive with equipment choices then... I have a superFlo VS (Without RS485) instead of an IntelliFlo 🤦‍♂️

**Reply by rstrouse** [reply]
No it is likely that we do not have all the messages coming off the ET8 for iChlor30 I will investigate the outputs this weekend.  

There is so much bad advice out there but here is a grain of salt for you.  Yes SuperFlo wasn't the best choice but it is workable but an IntelliChlor 40 cell would have been a better choice.

**Reply by VincentLombardi** [reply]
Ah okay gotcha! Thanks for taking a look. I really appreciate all the work you folks put into this project.

Yeah, before I discovered the TFP forums, this project, and other internet resources I received some really bad advice. The SuperFlo VS was purchased in a rush and before I even knew what the ET was capable of and just needed a pump. Slowly piecing everything together now but even my PB didn't know what he was doing when the pool was built... he wired up the ET control center transformer wrong, blew it up, and never told me. So the ET was never set up for the first 8 years of owning the pool. No temp sensors installed, no power to the mainboard, just manual control for everything, the ET just acted as an electrical subpanel.

Luckily, I'm an automation engineer by trade, so I've slowly been getting everything wired up and replaced. Converted my ET4 to an ET8 by soldering a resistor on the board, flashed the mainboard with new firmware, and re-doing the load center this summer to actually use the relays. 

Here is the enclosure I'm building for REM and some of my other goodies:
![IMG_2132](https://github.com/tagyoureit/nodejs-poolController/assets/33071399/974f0afc-9682-4e17-975a-286bf50347db)
![IMG_2133](https://github.com/tagyoureit/nodejs-poolController/assets/33071399/c93c241e-06e6-4624-be1c-466bfcd8edb0)

**Reply by rstrouse** [reply]
Heck I would ditch the ET8 at this point and go full on Nixie.  It is much more capable than ET.

**Reply by VincentLombardi** [reply]
I thought about it. Unfortunately I have already sunk so much money into the ET (new xformer, 5 extra relays, rs485 board, etc.) that I might as well just use it. Though, I am adding a lot of extra functionality through REM.

**Comment by VincentLombardi** [comment]
Hello!

Just wanted to follow up on this. Let me know if you need any replays from me or need me to test any functionality.

Thanks!

**Reply by rstrouse** [reply]
Yes can you generate a replay when you are changing the iChlor output from ET?

**Reply by VincentLombardi** [reply]
[OCP.Control.zip](https://github.com/tagyoureit/nodejs-poolController/files/12016865/OCP.Control.zip)

I went onto the OCP and adjusted the Pool Output %, waited a few seconds, and saw it update on the iChlor 30s screen. No update in NJSPC though.

**Reply by rstrouse** [reply]
Ok so I didn't look at the obvious.  Your iChlor was added with an External master.  This is used for chaining of multiple chlorinators.  You need to go into the Chemistry tab and delete it.  Then add it back in but this time tell it that the master is EasyTouch.

**Reply by VincentLombardi** [reply]
So I deleted it and tried adding a new chlorinator with the Master as EasyTouch 8P. With the Model set to unknown (since iChlor30 is not in the list) I get the following when clicking "save chlorinator"

![image](https://github.com/tagyoureit/nodejs-poolController/assets/33071399/783f7780-1d60-41b5-bd7a-7f406b155c82)

Here is the stack trace:

TypeError: Cannot read properties of undefined (reading 'val')
    at TouchChlorinatorCommands.setChlorAsync (/app/controller/boards/EasyTouchBoard.ts:1968:71)
    at /app/web/services/config/Config.ts:668:57
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at next (/app/node_modules/express/lib/router/route.js:144:13)
    at Route.dispatch (/app/node_modules/express/lib/router/route.js:114:3)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at /app/node_modules/express/lib/router/index.js:284:15
    at Function.process_params (/app/node_modules/express/lib/router/index.js:346:12)
    at next (/app/node_modules/express/lib/router/index.js:280:10)
    at /app/web/Server.ts:749:25
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/app/node_modules/express/lib/router/index.js:328:13)
    at /app/node_modules/express/lib/router/index.js:286:9
    at Function.process_params (/app/node_modules/express/lib/router/index.js:346:12)
    at next (/app/node_modules/express/lib/router/index.js:280:10)
    at /app/node_modules/body-parser/lib/read.js:137:5

If I set the model to say, IntelliChlor IC20, I get the same error with the same stack trace

**Reply by VincentLombardi** [reply]
The chlorinator was automatically added previously to NJSPC when it polled the config, and I changed the chlorination % on the OCP. Then I could modify the model and other parameters. Adding it manually throws the above error

**Reply by rstrouse** [reply]
Set the body to pool before you hit save.  The IChlor is not picking up the default body settings.

**Reply by VincentLombardi** [reply]
Tried that, but when I go to add a new chlorinator it already auto-populates the Pool as the body. I tried clicking the drop-down any clicking pool anyways but same error:

![image](https://github.com/tagyoureit/nodejs-poolController/assets/33071399/6f17226c-f4a4-4a04-8413-ff0a4e39c502)

![image](https://github.com/tagyoureit/nodejs-poolController/assets/33071399/eb043736-6e47-4a96-b299-c91b449265b3)

![image](https://github.com/tagyoureit/nodejs-poolController/assets/33071399/13dfb0cb-6a57-4cd7-b3c6-a38911aab5fd)

Here is a replay of my attempt to add it:

[replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/12017664/replay.zip)

**Comment by rstrouse** [comment]
Thanks for that.  Pull njsPC there was an issue with ET `P` models and the initial information fed for the chlorinator.  I also added iChlor 30 to the models.

**Reply by VincentLombardi** [reply]
Awesome! I use Docker compose with @emes's latest image [here](https://hub.docker.com/r/msmi/nodejs-poolcontroller). I will have to wait for the image to reflect the new changes and then I can give it a try.

Thanks for your help!
